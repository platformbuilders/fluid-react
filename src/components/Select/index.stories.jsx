import React from 'react';
import { text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import markdown from './select.md';
import Select from './index';

storiesOf('Select', module)
  .addDecorator(withKnobs)
  .addParameters({
    readme: {
      content: markdown,
    },
    viewport: {
      defaultViewport: '',
    },
  })
  .add('Example', () => (
    <Select
      style={{ width: 100 }}
      label="Gender"
      error={text('error', '')}
      selectedValue={text('selectedValue', '')}
      values={[
        {
          value: 'Male',
          label: 'Male',
        },
        {
          value: 'Female',
          label: 'Female',
        },
      ]}
    />
  ));
