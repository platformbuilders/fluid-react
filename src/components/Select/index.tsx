import { FC, SelectHTMLAttributes } from 'react';
import { ChevronDownIcon, ChevronUpIcon } from '@radix-ui/react-icons';
import {
  Icon,
  ScrollDownButton,
  ScrollUpButton,
  Value,
} from '@radix-ui/react-select';
import {
  PlaceholderText,
  StyledContent,
  StyledItem,
  StyledItemText,
  StyledTrigger,
  StyledViewPort,
  Wrapper,
} from './styles';

type SelectOptions = {
  option: string;
  value: string;
};

export type SelectProps = SelectHTMLAttributes<HTMLSelectElement> & {
  id: string;
  options: SelectOptions[];
  disabled?: boolean;
  helperMessage?: string;
  label?: string;
  placeholder?: string;
  onValueChange: (value: string) => void;
  defaultValue?: string;
  value?: string;
  style?: any;
  styleItem?: any;
  styleContentItem?: any;
};

const Select: FC<SelectProps> = ({
  options,
  disabled = false,
  placeholder,
  onValueChange,
  value,
  defaultValue,
  id,
  style,
  styleItem,
  styleContentItem,
}) => (
  <Wrapper
    onValueChange={onValueChange}
    defaultValue={defaultValue}
    value={value}
  >
    <StyledTrigger
      id={id}
      aria-label="select"
      disabled={disabled}
      style={style}
    >
      <Value placeholder={<PlaceholderText>{placeholder}</PlaceholderText>} />
      <Icon>
        <ChevronDownIcon />
      </Icon>
    </StyledTrigger>
    <StyledContent side="top" position="popper">
      <ScrollUpButton>
        <ChevronUpIcon />
      </ScrollUpButton>
      <StyledViewPort style={styleContentItem}>
        {options.map((item) => (
          <StyledItem value={item.value} key={item.value} style={styleItem}>
            <StyledItemText>{item.option}</StyledItemText>
          </StyledItem>
        ))}
      </StyledViewPort>
      <ScrollDownButton>
        <ChevronDownIcon />
      </ScrollDownButton>
    </StyledContent>
  </Wrapper>
);

export default Select;
