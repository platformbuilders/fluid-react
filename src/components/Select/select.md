
# Select 

<br/>
<!-- STORY -->

<br/>

### Source Code
<!-- SOURCE -->

<!-- PROPS -->


