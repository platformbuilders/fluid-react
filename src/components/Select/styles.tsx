import styled from 'styled-components';

export const Wrapper = styled.div``;

export const StyledSelect = styled.select`
  .MuiSelect-select {
    width: calc(100% - 32px) !important;
  }
`;

export const Item = styled.div`
  .MuiListItem-root {
    width: 100% !important;
  }
`;
