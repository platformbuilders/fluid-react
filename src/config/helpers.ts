import { DefaultTheme } from 'styled-components';
import { animations, colors, radius, spacing } from '../theme';

export const theme: DefaultTheme = {
  ...colors,
  ...animations,
  ...radius,
  ...spacing,
};
