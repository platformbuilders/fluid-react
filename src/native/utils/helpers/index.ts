export * from '@platformbuilders/helpers';
export * from './generateHaptic';
