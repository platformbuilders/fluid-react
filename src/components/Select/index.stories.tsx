import type { Meta, StoryObj } from '@storybook/react';

import Select from './index';

const meta: Meta<typeof Select> = {
  title: 'Select',
  component: Select,
  args: {
    defaultValue: 'Female',
    options: [
      {
        value: 'Male',
        option: 'Male',
      },
      {
        value: 'Female',
        option: 'Female',
      },
    ],
  },
};

export default meta;
type Story = StoryObj<typeof Select>;

export const Default: Story = {};
