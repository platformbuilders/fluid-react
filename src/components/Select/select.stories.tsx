import { expect, jest } from '@storybook/jest';
import type { Meta, StoryObj } from '@storybook/react';
import { configure, userEvent, within } from '@storybook/testing-library';
import Select from './index';

configure({ testIdAttribute: 'id' });

//  Mocks
const mockTextId = 'select-test-id';
const mockOnChange = jest.fn();

const meta: Meta<typeof Select> = {
  title: 'Components/Select',
  component: Select,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: {
    id: mockTextId,
    helperMessage: 'Select Test',
    label: 'Select Test',
    value: 'Female',
    style: {},
    styleContentItem: {},
    styleItem: {},
    options: [
      {
        value: 'Male',
        option: 'Male',
      },
      {
        value: 'Female',
        option: 'Female',
      },
    ],
    onValueChange: mockOnChange,
  },
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);

    await step('Select | Test Render', async () => {
      expect(canvas.getByTestId(mockTextId)).toBeInTheDocument();
    });

    await step('Radio | Select Male', async () => {
      await userEvent.click(canvas.getByTestId(mockTextId));
      await userEvent.click(canvas.getByText('Male'));
      expect(mockOnChange).toHaveBeenCalled();
    });
  },
};

type Story = StoryObj<typeof Select>;

export const Default: Story = {
  args: {},
};

export default meta;
