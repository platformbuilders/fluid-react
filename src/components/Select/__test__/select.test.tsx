import { ThemeProvider } from 'styled-components';
import { fireEvent, render } from '@testing-library/react';

import { theme } from '../../../config/helpers';
import Select, { SelectProps as Props } from '../index';

const defaultProps: Props = {
  label: 'Gender',
  selectedValue: 'Male',
  onChange: jest.fn(),
  values: [
    {
      value: 'Male',
      label: 'Male',
    },
    {
      value: 'Female',
      label: 'Female',
    },
  ],
};

describe('Component: Select', () => {
  test('snapshots with default props', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Select {...defaultProps} />
      </ThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('should change value to female', () => {
    // should
    const handleChange = jest.fn();

    // when
    const { getByRole } = render(
      <Select id="select-test" {...defaultProps} onChange={handleChange} />,
    );
    fireEvent.change(getByRole('textbox', { hidden: true }), {
      target: { value: 'Female' },
    });
    expect(handleChange).toHaveBeenCalled();
  });
});
