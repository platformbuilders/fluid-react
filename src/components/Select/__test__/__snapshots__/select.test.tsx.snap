// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Select snapshots with default props 1`] = `
<div>
  <div
    class="MuiFormControl-root sc-bczRLJ"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
    >
      Gender
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline sc-gsnTZi BhRRc MuiInputBase-formControl MuiInput-formControl"
    >
      <div
        aria-haspopup="listbox"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        role="button"
        tabindex="0"
      >
        Male
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        tabindex="-1"
        value="Male"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-filled"
    />
  </div>
</div>
`;
